### PR TITLE
Adding the Bitrix24 provider

### PR DESCRIPTION
**Pull request by it-delta**
OAuth2 Provider for Bitrix24
[bitrix24.com](url)

Docs: [here](https://training.bitrix24.com/rest_help/oauth/authentication.php)

### CODE REVIEW

File: src/Bitrix24/Bitrix24Provider.php [hunk 1]
```diff
@@ -0,0 +1,91 @@
+<?php
+
+namespace SocialiteProviders\Bitrix24;
+
+use SocialiteProviders\Manager\OAuth2\AbstractProvider;
+use SocialiteProviders\Manager\OAuth2\User;
+
+class Bitrix24Provider extends AbstractProvider
```
Comment (lucasmichot) [inline src/Bitrix24/Bitrix24Provider.php:8]: I would prefer to stay consistent with the other providers and name this class **Provider**

File: src/Bitrix24/Bitrix24Provider.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+{
+    /**
+     * Unique Provider Identifier.
+     */
+    const IDENTIFIER = 'BITRIX24';
```
Comment (lucasmichot) [inline src/Bitrix24/Bitrix24Provider.php:13]:
```suggestion
    public const IDENTIFIER = 'BITRIX24';
```

File: src/Bitrix24/Bitrix24Provider.php [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    protected $scopes = [''];
+
+    public static function additionalConfigKeys()
```
Comment (lucasmichot) [inline src/Bitrix24/Bitrix24Provider.php:20]: Missing `inheritdoc` docblock here

File: src/Bitrix24/Bitrix24Provider.php [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    /**
+     * Get the portal URL.
+     *
+     * @return string
+     */
+    protected function getPortalUrl()
+    {
+        return $this->getConfig('endpoint');
```
Comment (lucasmichot) [inline src/Bitrix24/Bitrix24Provider.php:40]: What is the default value if `BITRIX24_ENDPOINT` environment variable is not provided?

File: src/Bitrix24/Bitrix24Provider.php [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        return 'https://oauth.bitrix.info/oauth/token/';
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getUserByToken($token)
+    {
+        $response = $this->getHttpClient()->get($this->getPortalUrl().'/rest/user.current/', [
+            'query' => [
```
Comment (lucasmichot) [inline src/Bitrix24/Bitrix24Provider.php:57]: Can you use `\GuzzleHttp\RequestOptions::QUERY` constant here instead?

File: src/Bitrix24/Bitrix24Provider.php [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+                'auth' => $token,
+            ],
+        ]);
+
+        $user = json_decode($response->getBody(), true);
+        if (isset($user['error'])) {
+            throw new \Exception($user['error'].': '.$user['error_description'], 403);
```
Comment (lucasmichot) [inline src/Bitrix24/Bitrix24Provider.php:64]: I would throw a RuntimeException instead, and I would import it.

File: src/Bitrix24/Bitrix24Provider.php [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+            'id'    => $user['ID'],
+            'name'  => trim($user['NAME'].' '.$user['LAST_NAME']),
+            'email' => $user['EMAIL'],
+        ]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getTokenFields($code)
```
Comment (lucasmichot) [inline src/Bitrix24/Bitrix24Provider.php:85]: This method can be dropped

File: src/Bitrix24/README.md [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+## Installation & Basic Usage
+
+Please see the [Base Installation Guide](https://socialiteproviders.com/usage/), then follow the provider specific instructions below.
+
+### Add configuration to `config/services.php`
+
+```php
+'bitix24' => [
+      'endpoint' => env('BITRIX24_ENDPOINT'),
```
Comment (lucasmichot) [inline src/Bitrix24/README.md:15]:
```suggestion
      'endpoint' => env('BITRIX24_ENDPOINT_URI'),
```

File: src/Bitrix24/README.md [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+Configure the package's listener to listen for `SocialiteWasCalled` events.
+
+Add the event to your `listen[]` array in `app/Providers/EventServiceProvider`. See the [Base Installation Guide](https://socialiteproviders.com/usage/) for detailed instructions.
+
+```php
+protected $listen = [
+    \SocialiteProviders\Manager\SocialiteWasCalled::class => [
+        // ... other providers
+        \SocialiteProviders\Amazon\Bitrix24ExtendSocialite::class.'@handle',
```
Comment (lucasmichot) [inline src/Bitrix24/README.md:32]: Amazon?

**Review by lucasmichot — CHANGES_REQUESTED** [review]
Thanks for the PR
Can you address the review comments?

Also make sure to add this provider in `monorepo-builder.yml`

File: src/Bitrix24/Bitrix24Provider.php (hunk 1 above)
Comment (it-delta) [reply src/Bitrix24/Bitrix24Provider.php:8]: Done

File: src/Bitrix24/Bitrix24Provider.php (hunk 2 above)
Comment (it-delta) [reply src/Bitrix24/Bitrix24Provider.php:13]: done

File: src/Bitrix24/Bitrix24Provider.php (hunk 3 above)
Comment (it-delta) [reply src/Bitrix24/Bitrix24Provider.php:20]: done

File: src/Bitrix24/README.md (hunk 9 above)
Comment (it-delta) [reply src/Bitrix24/README.md:32]: done

File: src/Bitrix24/Bitrix24Provider.php (hunk 5 above)
Comment (it-delta) [reply src/Bitrix24/Bitrix24Provider.php:57]: done

File: src/Bitrix24/Bitrix24Provider.php (hunk 7 above)
Comment (it-delta) [reply src/Bitrix24/Bitrix24Provider.php:85]: ok

File: .idea/workspace.xml [hunk 10]
```diff
@@ -0,0 +1,71 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (lucasmichot) [inline .idea/workspace.xml:1]: Can you remove this file?

**Review by lucasmichot — APPROVED** [review]

**Review by atymic — APPROVED** [review]

**Pull request merged**